appscale/tools/node_layout.py

```diff
@@ -367,202 +317,259 @@ def is_valid_simple_format(self):
+                    for i in xrange(node_set.get('nodes'))]
+      # Otherwise get the ips and validate them.
+      else:
+        ip_or_ips = node_set.get('nodes')
+        ips_list = ip_or_ips if isinstance(ip_or_ips, list) else [ip_or_ips]
+        # Validate that the ips_list are either node-id or ip addresses.
+        for ip in ips_list:
+          self.is_cloud_ip(ip)
+
+      if len(ips_list) == 0:
+        self.invalid("Node amount cannot be zero.")
+
+      # Get the roles.
+      role_or_roles = node_set.get('roles')
+      roles = role_or_roles if isinstance(role_or_roles, list) else \
+        [role_or_roles]
+
+      # Immediately fail if we have more than one node for master.
+      if 'master' in roles and master_node and len(ips_list) > 1:
+        self.invalid("Only one master is allowed.")
+
+      # Create or retrieve the nodes from the node_hash.
+      nodes = [node_hash[ip] if ip in node_hash else AdvancedNode(ip, cloud)
+               for ip in ips_list]
+
+      # Validate volume usage, there should be an equal number of volumes to
+      # number of nodes.
+      if node_set.get('disks', None):
+        disks = node_set.get('disks')
+        self.validate_disks(nodes, disks)
+
+        for node, disk in zip(nodes, disks):
+          node.disk = disk
+
+      # Add the defined roles to the nodes.
       for node in nodes:
-        if node.is_role('login'):
-          node.public_ip = self.login_host
+        for role in roles:
+          node.add_role(role)
 
-    appengine_count = 0
-    for node in nodes:
-      if node.is_role('appengine'):
-        appengine_count += 1
+      # Check cases where a master is needed.
+      if 'master' in roles:
+        master_node = nodes[0]
+      if 'database' in roles and not db_master_created:
+        nodes[0].add_db_role(is_master=True)
+        db_master_created = True
+      if 'taskqueue' in roles and not tq_master_created:
+        # Check if we have more than one node to choose from and the first node
+        # is already the database master.
+        if 'db_master' in nodes[0].roles and len(nodes) > 1:
+          nodes[1].add_taskqueue_role(is_master=True)
+        else:
+          nodes[0].add_taskqueue_role(is_master=True)
+      if 'login' in roles:
+        login_found = True
+      elif 'login' in roles and login_found:
+        self.invalid("Only one login is allowed.")
 
-    if appengine_count < 1:
-      return self.invalid("Need to specify at least one appengine node")
+      # Update dictionary containing role counts.
+      role_count.update({(role, role_count.get(role, 0) + len(nodes))
+                         for role in roles})
+      # Update the node_hash with the modified nodes.
+      node_hash.update({(node.public_ip, node) for node in nodes})
 
-    memcache_count = 0
-    for node in nodes:
-      if node.is_role('memcache'):
-        memcache_count += 1
+      # Update node_count.
+      node_count += len(nodes)
```

Comment:
I think I should not allow being able to specify node-id since I'm using them for the integers. If you think about it there's really no need for people to specify node-1 since they'd be using it to put multiple roles and now its role based.